src/interfaces/records/handlers/records-query.ts

```diff
@@ -114,12 +129,17 @@ export class RecordsQueryHandler implements MethodHandler {
       ...RecordsQuery.convertFilter(recordsQuery.message.descriptor.filter),
       interface         : DwnInterfaceName.Records,
       method            : DwnMethodName.Write,
-      // TODO: `recordsQuery.author` cannot be undefined until #299 is implemented (https://github.com/TBD54566975/dwn-sdk-js/issues/299)
       recipient         : recordsQuery.author!,
       isLatestBaseState : true,
       published         : false
     };
-    const unpublishedRecordsForQueryAuthor = await StorageController.query(this.messageStore, this.dataStore, tenant, filter);
+    let unpublishedRecordsForQueryAuthor = await StorageController.query(this.messageStore, this.dataStore, tenant, filter);
+
+    // removing records that have its author being the same as the query author,
+    // because those records will be included in records returned by `fetchUnpublishedRecordsByAuthor()` already
+    // TODO: #392 - use a NOT query directly in the filter as a performance improvement (https://github.com/TBD54566975/dwn-sdk-js/issues/392)
+    unpublishedRecordsForQueryAuthor = unpublishedRecordsForQueryAuthor.filter(message => Message.getAuthor(message) !== recordsQuery.author);
```

Comment:
I have a slight preference for doing the dedupe in `fetchRecordsAsNonOwner` instead of here. By doing the dedupe here, we omit some of the records that conceptually fall into the category of "records where the query author is the `recipient`".

Unless there's a performance reason you're doing it here, I think it would be more conceptually clean to remove this `.filter()` and do `const records = [...publishedRecords, ...unpublishedRecordsByAuthor, ...unpublishedRecordsForQueryAuthor];` in `fetchRecordsAsNonOwner()` above.

Reply:
Agree the method name can now mislead even though I explicitly update documentation here as well as method summary to setup expectation. It is indeed for perf reason that I did this, which is why I went as far as creating another issue to improve the inefficiency further.

I went ahead and implemented the dedup logic though, since you are right that my approach is probably a premature optimization!!! I was originally not convinced that iterating the returned entries to compute their hash as map key is the "right" approach as it seems quite heavy in the name of de-dup when you could avoid it in the first place. But sneakily I used `recordId` as the key for now, it might not suffice when we return commits, but we can worry about it when we cross that bridge.